waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,381 @@
+        saml-response (-> form-params
+                        (get "SAMLResponse")
+                        base64->str
+                        xml-string->saml-resp)
+        assertions (.getAssertions saml-response)
+        _ (when-not (= 1 (count assertions))
+            (throw (ex-info (str "Could not authenticate user. Invalid SAML response. "
+                                 "Must have exactly one assertion but got " (count assertions))
+                            {:status 400})))
+        assertion (first assertions)
+        _ (validate-saml-assertion-signature assertion saml-signature-validator)
+        {:keys [attrs confirmation name-id]} (parse-saml-assertion assertion)
+        not-on-or-after (clj-time.coerce/from-sql-time (:not-on-or-after confirmation))
+        current-time (t/now)
+        _ (when-not (t/before? current-time not-on-or-after)
+            (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                            {:current-time current-time
+                             :expiry-time not-on-or-after
+                             :status 400})))
+        email (first (get attrs "email"))
+        ; https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/technical-reference/the-role-of-claims
+        upn (first (get attrs "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn"))
+        name-id-value (get name-id :value)
+        saml-principal (or upn name-id-value)
+        {:keys [authorization/principal authorization/user]} (auth/auth-params-map saml-principal)
+        saml-principal' (if (and email (= principal user)) email saml-principal)
+        saml-auth-data (utils/map->base-64-string
+                         {:not-on-or-after not-on-or-after :redirect-url request-url :saml-principal saml-principal'}
+                         password)]
+    {:body (render-authenticated-redirect-template {:auth-redirect-uri (str scheme "://" host auth-redirect-endpoint)
+                                                    :saml-auth-data saml-auth-data})
+     :status 200}))
+
+(defn saml-auth-redirect-handler
+  "Endpoint for POST back to Waiter with SAML authentication data. If data is still valid,
+   add waiter authentication cookie and redirect back to the original user app."
+  [{:keys [password]} request]
+  {:pre [(not-empty password)]}
+  (if-let [saml-auth-data (get-in (ring-params/params-request request) [:form-params "saml-auth-data"])]
+    (let [{:keys [not-on-or-after redirect-url saml-principal]}
+          (try
+            (utils/base-64-string->map saml-auth-data password)
+            (catch Exception e
+              (throw (ex-info "Could not parse saml-auth-data." {:inner-exception e
+                                                                 :saml-auth-data saml-auth-data
+                                                                 :status 400} e))))
+          current-time (t/now)
```

Comment:
Should we validate the values for: `not-on-or-after` `redirect-url` `saml-principal` before use?

Reply:
It would be nice to be more explicit about what went wrong in the 500 error we report.

Reply:
done